ie_serving/server/service.py

```diff
@@ -80,49 +82,25 @@ def Predict(self, request, context):
             logger.debug("PREDICT, problem with input data. Exit code {}"
                          .format(code))
             return predict_pb2.PredictResponse()
+
         target_engine = self.models[model_name].engines[version]
-        target_engine.in_use.acquire()
-        ################################################
-        # Reshape network inputs if needed
-        reshape_param = target_engine.detect_shapes_incompatibility(
-            inference_input)
-        if reshape_param is not None:
-            error_message = target_engine.reshape(reshape_param)
-            if error_message is not None:
-                code = statusCodes['invalid_arg'][GRPC]
-                context.set_code(code)
-                context.set_details(error_message)
-                target_engine.in_use.release()
-                return predict_pb2.PredictResponse()
-        ################################################
-        inference_start_time = datetime.datetime.now()
-        inference_output, error_message = target_engine.infer(
-            inference_input)
-        if error_message is not None:
+        inference_request = Request(inference_input)
+        target_engine.requests_queue.put(inference_request)
+        inference_output, used_ireq_index = inference_request.wait_for_result()
+        if inference_output is str:
             code = statusCodes['invalid_arg'][GRPC]
             context.set_code(code)
-            context.set_details(error_message)
-            target_engine.in_use.release()
+            context.set_details(inference_output)
+            logger.debug("PREDICT, problem during inference execution. Exit "
+                         "code {}".format(code))

```

Comment:
You need to release your `ireq_index` back to the queue in order to not lose it in case of error:
```python
target_engine.free_ireq_index_queue.put(used_ireq_index)
```